man/ip6.3
.TH qlibs: ipv6
.SH NAME
ipv6 \- IPv6 address evaluation and conversion
.SH SYNOPSIS 
.B #include \(dqip.h\(dq

unsigned int \fBip6_fmt\fP(char *\fIs\fR,char \fIip[16]\fR);
.br
unsigned int \fBip6_fmt_flat\fP(char *\fIs\fR,char \fIip[16]\fR);
.br
unsigned int \fBip6_scan_flat\fP(const char *\fIs\fR,char \fIip[16]\fR);
.br
unsigned int \fBip6_scan\fP(const char *\fIs\fR,char \fIip[16]\fR);
.br
unsigned int \fBip6_scanbracket\fP(const char *\fIs\fR,char \fIip[16]\fR);
.br
unsigned int \fBip6_ifscan\fP(char *\fIs\fR,char \fIip[16]\fR,stralloc *\fIifname\fR);
.br
unsigned int \fBip6_cidr\fP(char *\fIs\fR,char \fIip[16]\fR,unsigned long *\fIplen\fR);
.br
unsigned int \fBip6_bytestring\fP(stralloc *\fIip6string\fR,char \fIip[16]\fR,int \fIplen\fR);
.SH DESCRIPTION
.B ip6_fmt
reads the
.I char[16]
IPv6 address and returns a compactified hexadecimal IPv6 address string
.IR fe80::fefe .

.B ip6_fmt_flat
reads the
.I char[16]
IPv6 address and returns all hexadecimal IPv6 address labels as string 
.IR fe80:0000:....:fefe .

.B ip6_scan
reads a compactified IPv6 address string
.I fe80::fefe
and converts it to the
.I char[16]
IPv6 address.

.B ip6_scan_flat
reads an uncompressed IPv6 address als hexadecimal string
and returns it's
.I char[16]
IPv6 address.

.B ip6_scanbracket
reads a compactified IPv6 address string enclosed in brackets
.I [fe80::fefe]
removes the brackets and calls
.B ip6_scan
on the result.

.B ip6_ifscan
reads the compactified IPv6 address string appended with the
interface_name
.I fe80::fefe%eth0
returns 
.I ifname 
and calls 
.B ip6_scan
for the (stripped) IPv6 address.

.B ip6_cidr
reads the compactified CIDR IPv6 address string
.I fe80::fefe/64
determines the prefix as integer
.I plen
and calls
.BR ip6_scan .
If no prefix is identfied, it returns 128.

.B ip6_bytestring
reads the IPv6 address given as
.I char[16]
while returning the 'bytestring' representation
.I 1001001....
up to the given prefix length
.IR plen .

The macro
.B ipv6_v4mapped
reads the IPv6 addresses given as
.I char[16]
and returns
.I 0
in case the given IPv6 address is not 
a IPv4 mapped address.
.SH "RETURN CODES"
The
.B ip(6)*
programs return the number of bytes processed;
except for
.B ip6_bytestring
returning
.I 0
on success; others on failure. 
.SH "SEE ALSO"
ip4(3), 
socket_if(3)


CHANGELOG
fehQlibs-09:	First public release based on Qlibs as 
		a common project of Kai Peter and Erwin Hoffmann.
fehQlibs-10:	Added IPV6_V6ONLY setsockoptions call.
		Improved poll handling in iopause.c considering poll RC.
		Added building of shared libraries + updated docs.
fehQlibs-11:	Added ia4_fmt, ia6_fmt and dns_cname (for *qmail).
		Changed dns_transmit lookup time constants. 
fehQlibs-12:	dns_ip, dns_cname, dns_txt, dns_name and dns_mx
		return now the number of DNS answers received 
		unlike the number of bytes for the given output. 
		Added convenience routines for forthcoming s/qmail.
fehQlibs-12a:	Fixed missing DNS A lookup in dns_ip.c.
fehQlibs-12b:	Removed (one) obsolete return value check 
		for iopause in timeoutconn.c.
fehQlibs-12c:	Checking carefully all dnsstub return codes
		while using a coherent scheme. 
fehQlibs-13:	Bugs fixed: str_copy (wrong paranthesis setting)
		UI changed:	log() -> log_who()
		Added: man str(3)
fehQlibs-13a:	Fixed wrong input boundary checking for multiple
	DNSCACHE variables settings in dns_rcip.c. (Tx. J.W.)
fehQlibs-13b:	Added missing str_copyb() function in str.c.
fehQlibs-13c:	iopause it RC < 0 forces end of connection loop.
	DNS stub return codes straightend with djbdnscurve6-36.
fehQlibs-13d:	Fixed integration bugs not displaying IPv6 addresses
		(dns_ip.c, dns_mx.c).
		Wrong dependency check in Makefile for SHAREDLIBS (tx. Alan S.). 
fehQlibs-14:	Added 'const' for most input arguments + updated man:
		byte, logmsg, str, pathexec. Patch contributed by Alan S.
		Added byte_fill and case_upper function (potentially for SRS).
		Included CLFAGS and LDFLAGS macros in Makefile.


ip6.c
#include "fmt.h"
#include "byte.h"
#include "scan.h"
#include "ip.h"
#include "str.h"

/**
	@file ip6.c
	@author djb, fefe, feh
	@source ucspi-tcp, ucspi-tcp6
	@brief handling of IPv6 addresses
*/

/***
 @brief ip6_fmt
        convert IPv6 address to compactified IPv6 address string
 @param input:  IPv6 char array
        output: pointer to IPv6 address string
 @return int length of address (ok > 0)
 */

unsigned int ip6_fmt(char *s,char ip[16])
{
  unsigned int len;
  unsigned int i;
  unsigned int temp, temp0;
  unsigned int compressing;
  unsigned int compressed;
  int j;

  len = 0;
  compressing = 0;
  compressed = 0;

  for (j = 0; j < 16; j += 2) {
    if (j == 12 && ip6_isv4mapped(ip)) {
      len += ip4_fmt(s,ip+12);
      break;
    }

    temp = ((unsigned long) (unsigned char) ip[j] << 8) +
            (unsigned long) (unsigned char) ip[j+1];

    temp0 = 0;
    if (!compressing && j < 16)
      temp0 = ((unsigned long) (unsigned char) ip[j+2] << 8) +
               (unsigned long) (unsigned char) ip[j+3];

    if (temp == 0 && temp0 == 0 && !compressed) {
      if (!compressing) {
        compressing = 1;
        if (j == 0) {
          if (s) *s++ = ':';
          ++len;
        }
      }
    } else {
      if (compressing) {
        compressing = 0;
        ++compressed;
        if (s) *s++ = ':';
        ++len;
      }
      i = fmt_xlong(s,temp);
      len += i;
      if (s) s += i;
      if (j < 14) {
        if (s) *s++ = ':';
        ++len;
      }
    }
  }
  if (compressing) { *s++ = ':'; ++len; }

  return len;
}

/***
 @brief ip6_fmt_flat
        convert IPv6 address to IPv6 address string
 @param input:  IPv6 char array
        output: pointer to IPv6 address string
 @return int length of address (ok > 0)
 */

unsigned int ip6_fmt_flat(char *s,char ip[16])
{
  int i;
  for (i = 0; i < 16; i++) {
    *s++ = tohex((unsigned char)ip[i] >> 4);
    *s++ = tohex((unsigned char)ip[i] & 15);
  }
  return 32;
}

/***
 @brief ia6_fmt
        convert IPv6 address to inverse DNS nibble format
        1.2.3.4.5.6.7.8.9.a.b.c.d.e.f.1.2.3.4.5.6.7.8.9.a.b.c.d.e.f.ip6.arpa 
 @param input:  IPv6 char array
        output: pointer to IPv6 address string
 @return int length of address 
 */

unsigned int ia6_fmt(char *s,char ip[16])
{
  unsigned int i;
  unsigned int len;
  int j;

  static char data[] = "0123456789abcdef";
  len = 0;

  for (j = 15; j >= 0; j--) {
    i = fmt_str(s,&data[ip[j] & 0x0f]); len += i; if (s) s += i;
    i = fmt_str(s,"."); len += i; if (s) s += i;
    i = fmt_str(s,&data[ip[j] >> 4 & 0x0f]); len += i; if (s) s += i;
    i = fmt_str(s,"."); len += i; if (s) s += i;
  }
  i = fmt_str(s,"ip6.arpa."); len += i; if (s) s += i;

  return len;
}


/***
 @brief ip6_scan_flat
        convert IPv6 address string to IPv6 address array
 @param input:  pointer to IPv6 address string
        output: IPv6 char array
 @return int length of address (ok > 0)
 */

unsigned int ip6_scan_flat(const char *s,char ip[16])
{
  int i, tmp;

  for (i = 0; i < 16; i++) {
    tmp = fromhex(*s++);
    if (tmp < 0) return 0;
    ip[i] = tmp << 4;
    tmp = fromhex(*s++);
    if (tmp < 0) return 0;
    ip[i] += tmp;
  }
  return 32;
}

/***
 @brief ip6_scan
        parse compactified IPv6 address string and convert to IPv6 address array
 @param input:  pointer to IPv6 address string
        output: IPv6 char array
 @return int length of ip6_address/ip
 */

unsigned int ip6_scan(const char *s,char ip[16])
{
  unsigned int i, j;
  unsigned int len = 0;
  unsigned long u;

  char suffix[16];
  int prefixlen = 0;
  int suffixlen = 0;

 /* Always return IPv4 as IPv4-mapped IPv6 address */
  if ((i = ip4_scan(s,ip+12))) {	
    for (len = 0; len < 12; ++len) 
      ip[len] = V4mappedprefix[len];
    return i;
    if (byte_equal(ip+12,4,V4localnet)) {
      byte_copy(ip,16,V6localnet);
      return 16;
    }
  }
  byte_zero(ip,16);

  for (;;) {
    if (*s == ':') {
      len++;
      if (s[1] == ':') {        /* Found "::", skip to part 2 */
        s += 2; len++;
        break;
      }
      s++;
    }
    i = scan_xlong((char *)s,&u);
    if (!i) return 0;

    if (prefixlen == 12 && s[i] == '.') {
      /* the last 4 bytes may be written as IPv4 address */
      i = ip4_scan(s,ip+12);
      if (i) 
        return i+len;
      else
        return 0;
    }

    ip[prefixlen++] = (u >> 8);
    ip[prefixlen++] = (u & 255);
    s += i; len += i;
    if (prefixlen == 16) return len;
  }

/* part 2, after "::" */
  for (;;) {
    if (*s == ':') {
      if (suffixlen == 0) break;
      s++;
      len++;
    } else if (suffixlen != 0) break;

    i = scan_xlong((char *)s,&u);
    if (!i) {
      len--;
      break;
    }

    if (suffixlen + prefixlen <= 12 && s[i] == '.') {
      j = ip4_scan(s,suffix+suffixlen);
      if (j) {
        suffixlen += 4;
        len += j;
        break;
      } else
        prefixlen = 12 - suffixlen; /* make end-of-loop test true */
    }

    suffix[suffixlen++] = (u >> 8);
    suffix[suffixlen++] = (u & 255);
    s += i; len += i;
    if (prefixlen + suffixlen == 16) break;
  }

  for (i = 0; i < suffixlen; i++)
    ip[16 - suffixlen + i] = suffix[i];

  return len;
}

/***
 @brief ip6_scanbracket
        parse IPv6 string address enclosed in brackets
 @param input:  pointer to IPv6 address string
        output: IPv6 char array
 @return int length of ip_address (ok > 0)
 */

unsigned int ip6_scanbracket(const char *s,char ip[16])
{
  unsigned int len;

  if (*s != '[') return 0;
  len = ip6_scan(s + 1,ip);
  if (!len) return 0;
  if (s[len + 1] != ']') return 0;
  return len + 2;
}

/**
 @brief ip6_ifscan
        parse compactified IPv6 address string 
        concatinated with the interface name: fe80::1%eth0
 @param input:  pointer to IPv6 address string
        output: IPv6 char array, stralloc interface_name
 @return int length of ip6_address/ip
 */

unsigned int ip6_ifscan(char *s,char ip[16],stralloc *ifname)
{
  int i; 
  int j = 0;
  int k = 0; 
  if (!stralloc_copys(ifname,"0")) return 0;

  if ((j = str_chr(s,'%'))) {
    if ((i = str_chr(s+j+1,' '))) k = i;
    else if ((i = str_chr(s+j+1,'\n'))) k = i;
    else if ((i = str_chr(s+j+1,'\t'))) k = i;
    if (k) s[j+k+1] = '\0'; /* input might contain trailing chars */
    if (!stralloc_copys(ifname,s+j+1)) return 0;
    s[j] = 0;
  }
  if (!stralloc_0(ifname)) return 0;

  return ip6_scan(s,ip);
}

/**
 @brief ip6_cidr
        parse compactified IPv6 address string
        concatinated with the prefix length:  fe80::1/64
 @param input:  pointer to IPv6 address string
        output: IPv6 char array, long plen 
 @return int length of ip6_address/ip
 */

unsigned int ip6_cidr(char *s,char ip[16],unsigned long *plen)
{
  unsigned int j = 0;
  *plen = 128UL;

  j = str_chr(s,'/');
  if (s[j] == '/') {
     s[j] = 0;
     j = scan_ulong(s+j+1,plen);
  }
  return ip6_scan((const char *)s,ip);
}

/**
 @brief ip6_bytestring
        parse IPv6 address and represent as char string with length prefix
 @param input:  IPv6 char array, prefix length
        output: pointer to stralloc bit string;
 @return 0, if ok; 1, memory shortage
 */

unsigned int ip6_bytestring(stralloc *ipstring,char ip[16],int prefix)
{
  int i, j;
  unsigned char lowbyte, highbyte;

  ipstring->len = 0;
  if (!stralloc_copys(ipstring,"")) return 1;
  if (!stralloc_readyplus(ipstring,128)) return 1;

  for (i = 0; i < 16; i++) {
    lowbyte = (unsigned char) (ip[i]) & 0x0f;
    highbyte = (unsigned char) (ip[i] >> 4) & 0x0f;

    for (j = 3; j >= 0; j--) {
      if (highbyte & (1<<j)) {
        if (!stralloc_cats(ipstring,"1")) return 1;
      } else {
        if (!stralloc_cats(ipstring,"0")) return 1;
      }
      if (prefix == 0) {
        if (!stralloc_0(ipstring)) return 1;
        else return 0;
      }
      prefix--;
    }
    for (j = 3; j >= 0; j--) {
      if (lowbyte & (1<<j)) {
        if (!stralloc_cats(ipstring,"1")) return 1;
      } else {
        if (!stralloc_cats(ipstring,"0")) return 1;
      }
      if (prefix == 0) {
        if (!stralloc_0(ipstring)) return 1;
        return 0;
      }
      prefix--;
    }
  }
  if (!stralloc_0(ipstring)) return 1;

  return 0;
}
